In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

## Question 1

In [6]:
np.__version__

'1.21.6'

## Question 2

In [7]:
dataset = pd.read_csv('/content/drive/MyDrive/ml-zoomcamp/session 1/data.csv')

In [8]:
len(dataset)

11914

There are 11914 records in the data set

## Question 3

In [9]:
dataset.groupby(['Make'])['Make'].count().sort_values(ascending=False).head(3)

Make
Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

The most frequent car manufacturers are Chevrolet, Ford and Volskwagen

## Question 4

In [10]:
len(dataset['Model'][dataset['Make'] == 'Audi'].unique().tolist())

34

The number of unique Audi car models is 34

## Question 5

In [11]:
dataset.columns[dataset.isna().any()].tolist()

['Engine Fuel Type',
 'Engine HP',
 'Engine Cylinders',
 'Number of Doors',
 'Market Category']

There are 5 columns with null values

## Question 6

In [12]:
dataset['Engine Cylinders'].median()

6.0

In [13]:
dataset['Engine Cylinders'].mode()

0    4.0
dtype: float64

In [14]:
dataset_fillna = dataset['Engine Cylinders'].fillna(4)

In [15]:
dataset_fillna.median()

6.0

The median value is still the same

## Question 7

In [16]:
x = dataset[dataset['Make'] == 'Lotus'][['Engine HP','Engine Cylinders']].drop_duplicates(keep='first')

In [17]:
x

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [18]:
def vector_vector_multiplication(u, v):
    assert u.shape[0] == v.shape[0]
    
    n = u.shape[0]
    
    result = 0.0

    for i in range(n):
        result = result + u[i] * v[i]
    
    return result

In [19]:
def matrix_vector_multiplication(U, v):
    assert U.shape[1] == v.shape[0]
    
    num_rows = U.shape[0]
    
    result = np.zeros(num_rows)
    
    for i in range(num_rows):
        result[i] = vector_vector_multiplication(U[i], v)
    
    return result

In [20]:
def matrix_matrix_multiplication(U,V):
  assert U.shape[1] == V.shape[0]
  
  num_rows = U.shape[0]
  num_cols = V.shape[1]

  result = np.zeros((num_rows,num_cols))

  for i in range(num_cols):
    vi = V[:, i]
    Uvi = matrix_vector_multiplication(U, vi)
    result[:, i] = Uvi
  
  return result

In [27]:
x = x.to_numpy()
x

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [28]:
xt = np.transpose(x)
xt

array([[189., 218., 217., 350., 400., 276., 345., 257., 240.],
       [  4.,   4.,   4.,   8.,   6.,   6.,   6.,   4.,   4.]])

In [31]:
xtx = matrix_matrix_multiplication(x,xt)
xtx

array([[ 35737.,  41218.,  41029.,  66182.,  75624.,  52188.,  65229.,
         48589.,  45376.],
       [ 41218.,  47540.,  47322.,  76332.,  87224.,  60192.,  75234.,
         56042.,  52336.],
       [ 41029.,  47322.,  47105.,  75982.,  86824.,  59916.,  74889.,
         55785.,  52096.],
       [ 66182.,  76332.,  75982., 122564., 140048.,  96648., 120798.,
         89982.,  84032.],
       [ 75624.,  87224.,  86824., 140048., 160036., 110436., 138036.,
        102824.,  96024.],
       [ 52188.,  60192.,  59916.,  96648., 110436.,  76212.,  95256.,
         70956.,  66264.],
       [ 65229.,  75234.,  74889., 120798., 138036.,  95256., 119061.,
         88689.,  82824.],
       [ 48589.,  56042.,  55785.,  89982., 102824.,  70956.,  88689.,
         66065.,  61696.],
       [ 45376.,  52336.,  52096.,  84032.,  96024.,  66264.,  82824.,
         61696.,  57616.]])

In [34]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])
y

array([1100,  800,  750,  850, 1300, 1000, 1000, 1300,  800])

In [40]:

#matrix_vector_multiplication(matrix_matrix_multiplication(xt,np.linalg.inv(xtx)), y)